services/harmony/app/util/edl-api.ts

```diff
@@ -194,6 +190,98 @@ export async function verifyUserEula(context: RequestContext, username: string,
   };
 }
 
+// type of EDL query results
+type EdlResults = EdlUserEulaInfo | EdlGroupMembership;
+
+// Enum defining supported types of edl requests
+export enum edlQueryType {
+  EULA = 'EULA',
+  GROUP = 'GROUP',
+}
+
+/**
+ * Wrapper function for EDL queries.
+ * Designed to be used as the `fetchMethod` in an LRUCache.
+ *
+ * @param key - The cache key (usually a hash of query parameters)
+ * @param _sv - The "stale value" from LRUCache; unused in this implementation
+ * @param options - Contains `context` used to submit EDL query request
+ * @returns A Promise resolving to a EdlResults object from the appropriate source
+ */
+async function fetchFromEdl(
+  key: string,
+  _sv: EdlResults,
+  { context: fetchContext })
+  : Promise<EdlResults> {
+  const { type, context, username, eulaId } = fetchContext;
+  if (type === edlQueryType.EULA) {
+    return _verifyUserEula(context, username, eulaId);
+  } else if (type === edlQueryType.GROUP) {
+    return _getEdlGroupInformation(context, key);
+  } else {
+    throw new Error(`Invalid EDL query type: ${type}`);
+  }
+}
+
+// In memory cache for EDL query results.
+export const edlCache = new LRUCache<string, EdlResults>({
+  ttl: env.edlCacheTtl,
+  maxSize: env.edlCacheSize,
+  sizeCalculation: (result, _key): number => {
+    try {
+      // Serialize to JSON and measure byte size
+      const size = Buffer.byteLength(JSON.stringify(result), 'utf8');
+      theLogger.debug(`Cached EDL result size: ${size}`);
+      return size;
+    } catch {
+      // Fallback if serialization fails
+      theLogger.error('Failed to serialize EDL result in Cache');
+      return 1000;
+    }
+  },
+  fetchMethod: fetchFromEdl,
+});
+
+/**
+ * Retrieve the user EULA verification result.
+ * Perform the verification if cache miss.
+ *
+ * @param context - Information related to the user's request
+ * @param username - The EDL username
+ * @param eulaId - The id of the EULA (from the collection metadata)
+ * @returns A promise which resolves to info about whether the user has accepted a EULA,
+ * and if not, where they can go to accept it
+ */
+export async function verifyUserEula(context: RequestContext, username: string, eulaId: string)
+  : Promise<EdlUserEulaInfo> {
+  const type = edlQueryType.EULA;
+  const inputString = JSON.stringify({
+    username,
+    eulaId,
+  });
+
+  // Compute MD5 hash
+  const key = createHash('md5').update(inputString).digest('hex');
+  const result = await edlCache.fetch(key, { context: { type, context, username, eulaId } });
+  return result as EdlUserEulaInfo;
+}
+
+/**
+ * Retrieve the harmony relevant group information for a user with two keys isAdmin and isLogViewer.
```

Comment:
I believe we make use of 4 groups - are we caching the info for all 4?

Reply:
We are caching for all of them. This docstring is what's on the baseline. I will update it.